src/java.base/share/classes/java/lang/ScopedValue.java

```diff
@@ -526,7 +531,7 @@ public static <T> Carrier where(ScopedValue<T> key, T value) {
     }
 
     private ScopedValue() {
-        this.hash = generateKey();
+        this.hash = hashGenerator != null ? hashGenerator.getAsInt() : generateKey();
```

Comment:
The updated version looks much better. It might just a bit clearer to read hashGenerator once here (it doesn't matter of course but stable fields tend to attract a lot of questions).

Reply:
Something like this?
```suggestion
        final IntSupplier hashGen = hashGenerator;
        this.hash = hashGen != null ? hashGen.getAsInt() : generateKey();
```